src/transpileCodeblocks/compiler.ts

```diff
@@ -101,31 +100,57 @@ export class Compiler {
   }
 }
 
-function createCompilerHost(
-  compilerOptions: ts.CompilerOptions,
-  externalResolutions: CompilerSettings['externalResolutions']
-): ts.LanguageServiceHost &
+type ModifiedCompilerHost = ts.LanguageServiceHost &
   ts.ModuleResolutionHost &
   Required<Pick<ts.LanguageServiceHost, 'writeFile'>> & {
     setScriptFileNames(files: string[]): void;
-  } {
+  };
+
+function createCompilerHost(
+  compilerOptions: ts.CompilerOptions,
+  externalResolutions: CompilerSettings['externalResolutions']
+): ModifiedCompilerHost {
   const virtualFiles: Record<string, { contents: string; version: number }> =
     {};
   let scriptFileNames: string[] = [];
 
-  return {
+  let resolvedModules: Record<string, ts.ResolvedModuleFull | undefined> = {};
+
+  let cachedFiles: Record<string, string | undefined> = {};
+
+  const host: ModifiedCompilerHost = {
     ...ts.createCompilerHost(compilerOptions),
     getCompilationSettings() {
       return compilerOptions;
     },
     fileExists(fileName) {
-      // console.log('fileExists', fileName)
-      return !!virtualFiles[normalize(fileName)] || ts.sys.fileExists(fileName);
+      const result =
+        !!virtualFiles[normalize(fileName)] || ts.sys.fileExists(fileName);
+      return result;
     },
     readFile(fileName: string) {
-      // console.log('readFile', fileName)
-      const virtual = virtualFiles[normalize(fileName)];
-      return virtual ? virtual.contents : ts.sys.readFile(fileName);
+      const normalized = normalize(fileName);
+      const virtual = virtualFiles[normalized];
+
+      if (virtual) {
+        return virtual.contents;
+      }
+
+      if (normalized in cachedFiles) {
+        const cacheResult = cachedFiles[fileName];
+
+        if (cacheResult) {
+          return cacheResult;
+        }
+      }
+
+      const contents = ts.sys.readFile(fileName);
+
+      if (contents) {
+        cachedFiles[normalized] = contents;
+      }
```

Comment:
Did you verify that this doesn't cause problems in watch mode? 
If it does, we might need a flag to swap behaviours here.